File: printer/printer.go
package printer

import (
	"fmt"
	"mdj-diff/diff"
	"strings"

	"github.com/jedib0t/go-pretty/v6/table"
	"github.com/jedib0t/go-pretty/v6/text"
)

const WIDTH = 80

var EQUALS_LINE = strings.Repeat("=", WIDTH)

type printer interface {
	RenderTitle([]diff.TableChange)

	RenderChangesetHeader(name string, changeCount int)
	RenderEntityHeader(diff.TableChange)
	RenderTable(table.Writer)
}

// text

var _ printer = &textPrinter{}

type textPrinter struct{}

func (t *textPrinter) RenderTitle(changes []diff.TableChange) {
	fmt.Printf("%s\nMDJ Diff\n%s\n\n", EQUALS_LINE, EQUALS_LINE)
	fmt.Printf("Entities changed: %d\n\n", len(changes))
}

func (t *textPrinter) RenderEntityHeader(change diff.TableChange) {
	change.Type.SetColor()
	fmt.Println(EQUALS_LINE)
	fmt.Printf("%-*s%*s\n", WIDTH/2, change.Name, WIDTH/2, change.Type.Label())
	fmt.Println(EQUALS_LINE)
	fmt.Printf(text.EscapeReset)
}

func (t *textPrinter) RenderChangesetHeader(name string, changeCount int) {
	fmt.Printf("%s %d:\n", name, changeCount)
}
func (t *textPrinter) RenderTable(tw table.Writer) { tw.Render() }

// markdown

var _ printer = &markdownPrinter{}

type markdownPrinter struct{}

func (m *markdownPrinter) RenderEntityHeader(change diff.TableChange) {
	fmt.Printf("# ***%s*** (%s)\n", change.Name, change.Type.Label())
}

func (m *markdownPrinter) RenderTitle(changes []diff.TableChange) {
	fmt.Printf("# MDJ Diff\n\n")
	fmt.Printf("*Entities changed: %d*\n\n", len(changes))
}

func (m *markdownPrinter) RenderChangesetHeader(name string, changeCount int) {
	fmt.Printf("### %s (%d)\n", name, changeCount)
}

func (m *markdownPrinter) RenderTable(tw table.Writer) { tw.RenderMarkdown() }

File: src/printer/printer.rs
use crate::{diff::diff::TableChange, types::Project};

// const WIDTH: usize = 80;
const EQUALS_LINE: &str =
    "================================================================================";

pub trait Printer {
    fn render_title(&self, changes: &Vec<TableChange>);

    fn render_changeset_header(name: String, change_count: usize);

    fn render_entity_header(&self, tc: &TableChange);

    fn render_table(&self, a: &Project, b: &Project, changes: &Vec<TableChange>);

    fn print(&self, a: &Project, b: &Project, changes: &Vec<TableChange>);
}

pub struct TextPrinter {}

impl Printer for TextPrinter {
    fn render_title(&self, changes: &Vec<TableChange>) {
        print!("{}\nMDJ Diff\n{}\n\n", EQUALS_LINE, EQUALS_LINE);
        print!("Entities changed: {}\n\n", changes.len());
    }

    fn render_changeset_header(name: String, change_count: usize) {
        print!("{} {}:\n", name, change_count)
    }

    fn render_entity_header(&self, change: &TableChange) {
        let mut o = String::new();
        o += EQUALS_LINE;
        o += &format!("{} - {}\n", change.name, change.change_type.to_string());
        o += EQUALS_LINE;

        change.change_type.print_color(&o);
    }

    fn render_table(&self, _a: &Project, _b: &Project, _changes: &Vec<TableChange>) {}

    fn print(&self, a: &Project, b: &Project, changes: &Vec<TableChange>) {
        self.render_title(changes);

        for change in changes.iter() {
            self.render_entity_header(change);
            // 	printTableAttributes(p, change)

            println!();

            if change.columns.len() > 0 {
                TextPrinter::render_changeset_header(String::from("Columns"), change.columns.len());
            }

            // 	for _, column := range change.Columns {
            // 		printColumnChanges(p, column)
            // 		fmt.Println()
            // 	}

            // 	if len(change.Relationships) > 0 {
            // 		p.RenderChangesetHeader("Relationships", len(change.Relationships))
            // 	}

            // 	for _, rel := range change.Relationships {
            // 		printRelationshipChanges(p, A, B, rel)
            // 		fmt.Println()
            // 	}

            // 	if len(change.Tags) > 0 {
            // 		p.RenderChangesetHeader("Tags", len(change.Tags))
            // 	}

            // 	for _, tag := range change.Tags {
            // 		renderChanges(p, tag.Name, tag.Changes)
            // 		fmt.Println()
            // 	}
        }
    }
}
